utils/polus-filepattern-util/docs/source/Examples.rst

```diff
@@ -0,0 +1,307 @@
+    is roughly equivalent to the y-axis in plotting/graphing notation.
+
+The approach of looking at the files and coming up with our own ``filepattern``
+generally works for small data sets. However, if you have a large data set with
+hundreds or thousands of files and want to get a first guess on a
+``filepattern`` that would account for all file names, you can use the built-in
+``infer_pattern`` function.
+
+.. code-block:: python
+
+    import filepattern, pathlib
+
+    filepath = pathlib.Path(__file__).parent
+    filepath = filepath.joinpath('data/Small_Fluorescent_Test_Dataset/image-tiles/')
+    files = list(filepath.iterdir())
+
+    pattern_guess = filepattern.infer_pattern(files)
+
+    print('Inferred Pattern: {}'.format(pattern_guess))
+
+.. code-block:: bash
+
+    Inferred Pattern: img_r00{r}_c00{t}.tif
+
+The ``infer_pattern`` function looks at all of the files you supply to it, and
+then comes up with a good guess at what a filepattern should look like, but it
+likely won't get the variable names correct (if the variable names matter).
+
+The above notation with ``{r}`` indicates that the ``r`` variable can only ever
+take on value of 0-9 because it is only permitted one space within the filename.
+If you wanted to capture two spaces for the variables, you would do ``{rr}`` so
+that the above pattern would become ``img_r0{rr}_c00{c}.tif``, or 
+``img_r0{rr}_c0{cc}.tif``. Alternatively, a variable length variable definition
+could be defined using the ``+`` notation, which would make the pattern
+``img_r{r+}_c{c+}.tif``. This will capture all numbers following ``img_r`` but
+occur before ``_c``. This is useful in cases where the file names are not all
+fixed with, so that you can parse files like ``img1.tif`` and ``img10.tif``
+using the same ``filepattern``.
+
+This fixed width variable definition was created for instances where filenaming
+conventions are completely numeric, as in the following case:
+
+.. code-block:: bash
+
+    001004.tif
+    005005.tif
+    002001.tif
+    002005.tif
+    ...
+
+In this case, a filepattern like ``{r+}.tif`` would capture the entire sequence
+of numbers. Instead, we may want to store the first three numbers in one
+variable and the second three numbers in a second variables: ``{yyy}{xxx}.tif``.
+
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Using the FilePattern Class
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Now that we have a ``filepattern``, we will create a ``FilePattern`` object to
+help us handle the data that gets parsed from the ``filepattern``. A
+``FilePattern`` object is also an iterable that allows us to cleverly loop over
+the data
+
+.. code-block:: python
+
+    import filepattern, pathlib, pprint
+
+    filepath = pathlib.Path(__file__).parent
+    filepath = filepath.joinpath('data/Small_Fluorescent_Test_Dataset/image-tiles/')
+
+    pattern = 'img_r00{y}_c00{x}.tif'
+
+    fp = filepattern.FilePattern(filepath,pattern)
+
+    for file in fp():
```

Comment:
your usage pattern is different here than the Md pattern where you call `iterate()`. has that one not been updated?

Reply:
Also , if the only thing the `__call__` function does is set groupby and kwargs, then doesn't it make sense for you to call it without the parenthesis which will make the usage cleaner? so `for file in fp:`?

Reply:
I could have it called both ways. The reason to use `__call__` is to define a custom `group_by` and matching variable values. However, I think it also makes sense to be able to iterate through values without defining anything so that `for file in fp` works the same as `for file in fp()`. I have changed the code accordingly.

Reply:
Also, the `iterate()` was outdated. That was the old code.